### ISSUE TITLE
Age model fails with error

### ISSUE THREAD

**Issue by Qwasser** [beniz/deepdetect#39]
I am trying to use model from https://github.com/GilLevi/AgeGenderDeepLearning
Everything starts normally but when I send request for prediction it fails with error:

```
INFO - Ignoring source layer data
dede: /usr/include/boost/smart_ptr/shared_ptr.hpp:653: typename boost::detail::sp_member_access<T>::type boost::shared_ptr<T>::operator->() const [with T = caffe::MemoryDataLayer<float>; typename boost::detail::sp_member_access<T>::type = caffe::MemoryDataLayer<float>*]: Assertion `px != 0' failed.
INFO - Ignoring source layer lossAborted (core dumped)
```

**Comment by beniz** [comment]
Let me understand, you are trying to use the Caffe model from gillevi with dd, is that correct ?

I should be able to help, but more inflation is needed: all prototxt files typically and service + train API calls so that I can reproduce.

**Comment by Qwasser** [comment]
Yes, I am trying to use pretrained caffe model. It can be downloaded from here http://www.openu.ac.il/home/hassner/projects/cnn_agegender/cnn_age_gender_models_and_data.0.0.2.zip
To use the model I created "models/age" folder and put there deploy.prototxt (renamed deploy_age) and age.caffemodel files from the archive. Then I had created service using following code:

``` python
model = {'repository': '/home/ubuntu/models/age''}
parameters_input = {'connector':'image',
                            'width': 256,
                            'height': 256}
parameters_mllib = {'nclasses': 8}
parameters_output = {}

dd.put_service("age",
                       model,
                       "age classification",
                       'caffe',
                       parameters_input,
                       parameters_mllib,
                       parameters_output)
```

and the query is
```
curl -X POST "http://localhost:8080/predict" -d "{\"service\":\"age\",\"parameters\":{\"input\":{\"width\":256,\"height\":256},\"output\":{\"best\":3}},\"data\":[\"http://i.ytimg.com/vi/0vxOhd4qlnA/maxresdefault.jpg\"]}"

```

I think the problem is in deploy.prototxt and it should be converted somehow

**Comment by beniz** [comment]
Yes you need a memorydata layer at the top, this is so dd can pass data to the net. Look at the deploy file from templates/caffe/googlenet as an example. I all try to reproduce as well on my spare time, but please let me know how it goes.

**Comment by Qwasser** [comment]
Added following text to deploy.prototxt:

```
layers  {
  name: "caffenet"
  type: MEMORY_DATA
  top: "data"
  top: "label"
  memory_data_param {
    batch_size: 32
    channels: 3
    height: 256
    width: 256
  }
}
```

And everything worked :) Should it be in the documentation?

**Comment by beniz** [comment]
Ah got it to work in parallel to you. The key is to convert the Caffe model from the old format to the new one, and then update the deploy.prototxt file as well.

I've put a model tarball for the 8 classes age model here: http://www.deepdetect.com/models/
It is ready to use, though I can't report on the accuracy.

It could be added to the 'model' page http://www.deepdetect.com/applications/model/

**Comment by Qwasser** [comment]
Have you run into a trouble mean.binaryproto has incorrect dimensions?

**Comment by beniz** [comment]
I believe dimension is 227x227.

**Comment by beniz** [comment]
Btw, the predict call does not yet read the `mean.binaryproto file`. I will commit this shortly. From my short tests, it does affect the results significantly on this dataset.

**Comment by beniz** [comment]
Do you know what the age classes correspond to by any chance ? (0 to what range, etc..). Thanks.

**Comment by Qwasser** [comment]
Yes, I just wanted to send them to you

**Comment by beniz** [comment]
OK good! I was just surprised that the example they give in their tarball, `example_image.jpg` seems to be reported as `60+` (0.53) and `0_2` (0.35). My own picture however is unfortunately not reported as younger...

**Comment by Qwasser** [comment]
Have you tried it in notebook from archive? Works fine there (even without mean substraction).

**Comment by beniz** [comment]
Somehow it works better for me without the mean substraction. I've investigated and failed to understand why. It's possible the `mean.binaryproto` is not in the exact same format as the ones produced by the Caffe `compute_mean`. The Caffe `classification.cpp` example uses a planar 32-bit float BGR or grayscale format.
Closing this since it appears to be working fine with dd anyways, can be reopened if the mean file issue is a problem in practice.
